packages/hardhat/contracts/minters/PaymentSplitterWithFeeUpgradeable.sol

```diff
@@ -0,0 +1,166 @@
+        uint256 serviceFeeBasisPoints_
+    ) internal initializer {
+        require(payees.length == shares_.length, "PaymentSplitter: payees and shares length mismatch");
+        require(payees.length > 0, "PaymentSplitter: no payees");
+
+        for (uint256 i = 0; i < payees.length; i++) {
+            _addPayee(payees[i], shares_[i]);
+        }
+
+        serviceFeeAddress = serviceFeeAddress_;
+        serviceFeeBasisPoints = serviceFeeBasisPoints_;
+    }
+
+    /**
+     * @dev The Ether received will be logged with {PaymentReceived} events. Note that these events are not fully
+     * reliable: it's possible for a contract to receive Ether without triggering this function. This only affects the
+     * reliability of the events, and not the actual splitting of Ether.
+     *
+     * To learn more about this see the Solidity documentation for
+     * https://solidity.readthedocs.io/en/latest/contracts.html#fallback-function[fallback
+     * functions].
+     */
+    receive() external payable virtual {
+        emit PaymentReceived(_msgSender(), msg.value);
+    }
+
+    /**
+     * @dev Getter for the total shares held by payees.
+     */
+    function totalShares() public view returns (uint256) {
+        return _totalShares;
+    }
+
+    /**
+     * @dev Getter for the total amount of Ether already released.
+     */
+    function totalReleased() public view returns (uint256) {
+        return _totalReleased;
+    }
+
+    /**
+     * @dev Getter for the amount of shares held by an account.
+     */
+    function shares(address account) public view returns (uint256) {
+        return _shares[account];
+    }
+
+    /**
+     * @dev Getter for the amount of Ether already released to a payee.
+     */
+    function released(address account) public view returns (uint256) {
+        return _released[account];
+    }
+
+    /**
+     * @dev Getter for the address of the payee number `index`.
+     */
+    function payee(uint256 index) public view returns (address) {
+        return _payees[index];
+    }
+
+    /**
+     * @dev Triggers a transfer to `account` of the amount of Ether they are owed, according to their percentage of the
+     * total shares and their previous withdrawals.
+     */
+    function release(address payable account) public virtual {
+        require(_shares[account] > 0, "PaymentSplitter: account has no shares");
+
+        uint256 totalReceived = address(this).balance + _totalReleased;
+        uint256 payment = (totalReceived * _shares[account]) / _totalShares - _released[account];
+
+        require(payment != 0, "PaymentSplitter: account is not due payment");
+
+        _released[account] = _released[account] + payment;
+        _totalReleased = _totalReleased + payment;
+
+        _sendPaymentWithFee(account, payment);
+    }
+
+    function _sendPaymentWithFee(address payable account, uint256 payment) private {
+        uint256 feeAmount = (payment * serviceFeeBasisPoints) / 10000;
```

Comment:
seeing this line made me want to add enforcement that `serviceFeeBasisPoints` cannot ever exceed 10K on accident. what's your take on this?

Reply:
@eladmallel sure.
I guess the best place would be when setting it, to require it's <= 10000 (instead of checking on each release)
I'll add that